File: 12-Trees/01-Binary-Tree-Traversals/cpp/level_order.cpp
/*
 * Problem: Binary Tree Level Order Traversal
 * Source:  LeetCode 102
 * Time: O(n), Space: O(n)
 * Tags: tree, bfs, traversal
 */
#include <bits/stdc++.h>
using namespace std;
struct TreeNode { int val; TreeNode *left, *right; TreeNode(int x):val(x),left(nullptr),right(nullptr){} };

vector<vector<int>> levelOrder(TreeNode* root) {
    vector<vector<int>> result;
    if (!root) return result;
    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        int sz = q.size();
        vector<int> level;
        while (sz--) {
            auto node = q.front(); q.pop();
            level.push_back(node->val);
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
        result.push_back(level);
    }
    return result;
}

int main() {
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    auto res = levelOrder(root);
    for (auto& level : res) { for (int x : level) cout << x << " "; cout << endl; }
    // 3 | 9 20 | 15 7
    return 0;
}

File: 12-Trees/01-Binary-Tree-Traversals/java/LevelOrder.java
import java.util.*;
public class LevelOrder {
    static class TreeNode { int val; TreeNode left, right; TreeNode(int x) { val = x; } }
    public static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;
        Queue<TreeNode> q = new LinkedList<>(); q.add(root);
        while (!q.isEmpty()) {
            int sz = q.size(); List<Integer> level = new ArrayList<>();
            while (sz-- > 0) {
                TreeNode node = q.poll(); level.add(node.val);
                if (node.left != null) q.add(node.left);
                if (node.right != null) q.add(node.right);
            }
            result.add(level);
        }
        return result;
    }
    public static void main(String[] args) {
        TreeNode root = new TreeNode(3); root.left = new TreeNode(9); root.right = new TreeNode(20);
        System.out.println(levelOrder(root));
    }
}
